macros/edr/data_monitoring/data_monitors_configuration/get_buckets_configuration.sql

```diff
@@ -41,24 +71,54 @@
+            from bucket_times
+        )
         select
             case
-                when last_run is null then min_start
-                when last_run < min_start then min_start
-                when last_run < backfill_start then last_run
-                else backfill_start
-            end as min_start
-        from min_times
+                {# This prevents gaps in buckets for the metric #}
+                when last_max_bucket_end is null then days_back_start {# When this is the first run of this metric #}
+                when last_max_bucket_end < days_back_start then days_back_start {# When the metric was not collected for a period longer than days_back #}
+                when last_min_bucket_end > days_back_start then days_back_start {# When the metric was collected recently, but for a period that is smaller than days_back #}
+                when last_max_bucket_end < backfill_start then last_max_bucket_end {# When the metric was not collected for a period longer than backfill_days #}
+                else {{ elementary.edr_timeadd(metric_properties.time_bucket.period, 'periods_to_backfill', 'last_max_bucket_end') }} {# When backfill reduce full time buckets from last_max_bucket_end to backfill #}
+            end as min_bucket_start,
+            case
+                {# This makes sure we collect only full bucket #}
+                when last_max_bucket_end is null or last_max_bucket_end < days_back_start or last_min_bucket_end > days_back_start
+                then {{ elementary.edr_timeadd(metric_properties.time_bucket.period, 'periods_until_max', 'days_back_start') }} {# Add full buckets to days_back_start #}
+                else {{ elementary.edr_timeadd(metric_properties.time_bucket.period, 'periods_until_max', 'last_max_bucket_end') }} {# Add full buckets to last_max_bucket_end #}
+            end as max_bucket_end
+        from full_buckets_calc
     {%- endset %}
 
-    {%- set min_bucket_start_query_result = elementary.result_value(min_bucket_start_query) %}
+    {# We assume we should also cosider sources as incremental #}
+    {% if not (elementary.is_incremental_model(elementary.get_model_graph_node(model), source_included=true) or unit_test) %}
+        {%- set buckets = elementary.agate_to_dicts(run_query(regular_bucket_times_query))[0] %}
+    {%- else %}
+        {%- set buckets = elementary.agate_to_dicts(run_query(incremental_bucket_times_query))[0] %}
+    {% endif %}
 
-    {%- if min_bucket_start_query_result %}
-        {{ return(min_bucket_start_query_result) }}
+    {%- if buckets %}
+        {%- set min_bucket_start = elementary.edr_quote(buckets.get('min_bucket_start')) %}
+        {%- set max_bucket_end = elementary.edr_quote(buckets.get('max_bucket_end')) %}
+        {{ return([min_bucket_start, max_bucket_end]) }}
     {%- else %}
-        {{ return(min_bucket_start) }}
+        {{ exceptions.raise_compiler_error("Failed to calc test buckets min and max") }}
```

Comment:
Breaking this part out to a separate macro would make the get_test_min_and_max macro less intimidating (up to you if we bundle it in this PR or not)